### ISSUE TITLE
K8s Scheduler crashes

### ISSUE THREAD

**Issue by FaKod** [argoproj/argo#656]
What did you do? 
Deployed Argo on K8s 1.8.5 and the hello-world example

What did you see? 
kube-scheduler crashes for some reason (See the exception below)

What did you expect?
a running example

K8s Version: 1.8.5
Argo Version: v2.0.0-alpha3

Once I've deployed an example with
```
argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml
```

The Kube Scheduler crashes with

```
I0105 05:57:33.386632       1 feature_gate.go:156] feature gates: map[PodPriority:true APIListChunking:true AllAlpha:true EnableEquivalenceClassCache:true TaintBasedEvictions:true APIResponseCompression:true KubeletConfigFile:true ServiceNodeExclusion:true DevicePlugins:true DynamicKubeletConfig:true TaintNodesByCondition:true CPUManager:true DebugContainers:true SupportIPVSProxyMode:true PersistentLocalVolumes:true HugePages:true ExpandPersistentVolumes:true LocalStorageCapacityIsolation:true ExperimentalCriticalPodAnnotation:true CustomResourceValidation:true Initializers:true MountPropagation:true RotateKubeletServerCertificate:true Accelerators:true]
I0105 05:57:33.416814       1 factory.go:706] Created equivalence class cache
I0105 05:57:34.117997       1 controller_utils.go:1041] Waiting for caches to sync for scheduler controller
I0105 05:57:34.218857       1 controller_utils.go:1048] Caches are synced for scheduler controller
I0105 05:57:34.219099       1 leaderelection.go:174] attempting to acquire leader lease...
I0105 05:57:51.549190       1 leaderelection.go:184] successfully acquired lease kube-system/kube-scheduler
I0105 05:57:51.549663       1 event.go:218] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"kube-system", Name:"kube-scheduler", UID:"c8244354-f140-11e7-aec6-06cb78000678", APIVersion:"v1", ResourceVersion:"162543", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' k8s-bc-18-master-1 became leader
E0105 05:57:51.650336       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/utils.go:77
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/equivalence_cache.go:198
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/generic_scheduler.go:316
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/generic_scheduler.go:258
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x30a0d89]

goroutine 318 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x404e860, 0x9a915f0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.GetEquivalencePod(0xc42129c948, 0x4641a9, 0xc420210500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/utils.go:77 +0x99
k8s.io/kubernetes/plugin/pkg/scheduler/core.(*EquivalenceCache).getHashEquivalencePod(0xc420a01b00, 0xc42129c948, 0x411a42, 0xc420b956e0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/equivalence_cache.go:198 +0x38
k8s.io/kubernetes/plugin/pkg/scheduler/core.podFitsOnNode(0xc42129c948, 0x9adfc60, 0xc420a45f80, 0xc42088b600, 0xc420a01a70, 0xc420a01b00, 0x8, 0x0, 0xc420857f40, 0x406f55, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/generic_scheduler.go:316 +0x602
k8s.io/kubernetes/plugin/pkg/scheduler/core.findNodesThatFit.func1(0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/core/generic_scheduler.go:258 +0x159
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize.func1(0xc42014c180, 0xc421492280, 0xc4202a7e00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47 +0xcb
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:49 +0x12c
```

When I delete the example workflow, everything is fine again...

Any idea what could be the cause?

**Comment by FaKod** [comment]
seems that adding ```controller: true``` to the Pods metadata solves the issue. F.e.:

```
...
ownerReferences:
  - apiVersion: workflows.argoproj.io
    blockOwnerDeletion: true
    controller: true
    kind: Workflow
    name: hello-world-fwr28
    uid: e0f89c23-f1dc-11e7-9bef-06b1d000076b
...
```

K8s 1.8.6 seems to have reworked the 1.8.5 code at https://github.com/kubernetes/kubernetes/blob/v1.8.5/plugin/pkg/scheduler/algorithm/predicates/utils.go#L77 
So my issue may be solved with 1.8.6 (that I cannot verify currently)

**Comment by jessesuen** [comment]
Thanks for all the legwork on this issue. I don't have easy access to 1.8.5 or 1.8.6 K8s at the moment to reproduce/test this. I'm unclear why `controller: true` would be required. Cascaded deletion seems to be working fine on 1.8.4 and the API describes it as optional:
```
	// If true, this reference points to the managing controller.
	// +optional
	Controller *bool `json:"controller,omitempty" protobuf:"varint,6,opt,name=controller"`
```

Let me shoot a question over to sig-api-machinery to check if `controller: true` expected to be set or if this would be a k8s bug. It's possible that both are true.

**Comment by jessesuen** [comment]
I found a region in GKE that supports 1.8.5 and installed it to test, but I am unable to reproduce this. My hello-world ran just fine:

```
$ kubectl get pod -o yaml hello-world-x2hjc -o json | jq '.metadata.ownerReferences'
[
  {
    "apiVersion": "workflows.argoproj.io",
    "blockOwnerDeletion": true,
    "kind": "Workflow",
    "name": "hello-world-x2hjc",
    "uid": "27809522-f1f0-11e7-a839-42010a8a0175"
  }
]

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.5", GitCommit:"cce11c6a185279d037023e02ac5249e14daa22bf", GitTreeState:"clean", BuildDate:"2017-12-07T18:09:00Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.5-gke.0", GitCommit:"2c2a807131fa8708abc92f3513fe167126c8cce5", GitTreeState:"clean", BuildDate:"2017-12-19T20:05:45Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

Granted, GKE appears to build their own version of Kubernetes so it's possible they are running different codebase/scheduler which may not be exposed to this.

**Comment by FaKod** [comment]
That's what I'm using:

```
> kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-16T03:16:50Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.5+coreos.0", GitCommit:"b8e596026feda7b97f4337b115d1a9a250afa8ac", GitTreeState:"clean", BuildDate:"2017-12-12T11:01:08Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by jessesuen** [comment]
I'm pretty sure this is a k8s bug fixed in the following commit:
https://github.com/kubernetes/kubernetes/commit/e4055c0df23abce921bc5f1d44a3599f28b64d1e
This:
```
if *ref.Controller {
```
became:
```
if ref.Controller != nil && *ref.Controller {
```

However, the problematic code seems to have been this way for a while so I don't know why it doesn't happen all the time, unless there's something special about the coreos distribution that somehow tickles this code path.

**Comment by jessesuen** [comment]
Here is the design proposal on ControllerRefs:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/controller-ref.md

The recommendation I received is to use the `NewControllerRef` function in `k8s.io/apimachinery/pkg/apis/meta/v1`. This function automatically sets `controller: true` which appears to be always desired when setting the ownership, from what I can tell.

**Comment by jessesuen** [comment]
For posterity. Here are scenarios in which you may not want to set `controller: true`, but set the other fields. Answers from sig-api-machinery:

* You have one controller managing scale of a thing and another controller managing the podspec of the thing.  Any controller creating a deployment, replication controller, or replicaset hits this with HPA
* Also cases around keeping content of a secret in sync with something via a controller, but the controller doesn't own the lifecycle of the thing.
* not all controllers are related to resources.  You can imagine controllers driven by external config.  Basically, the controller flag and the 'only one of' semantic doesn't reflect a general pattern in keeping with the meta API

**Comment by FaKod** [comment]
I managed to test K8s v1.8.6+coreos.0 with argo v2.0.0-alpha3
Seems that the kube-scheduler crash is actually fixed with 1.8.6

**Comment by jessesuen** [comment]
Good to know. Thanks!
